src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -918,7 +918,10 @@ impl LinkCollector<'_, '_> {
             for md_link in preprocessed_markdown_links(&doc) {
                 let link = self.resolve_link(item, item_id, module_id, &doc, &md_link);
                 if let Some(link) = link {
-                    self.cx.cache.intra_doc_links.entry(item.item_id).or_default().push(link);
+                    let entry = self.cx.cache.intra_doc_links.entry(item.item_id).or_default();
+                    if entry.iter().find(|other| **other == link).is_none() {
```

Comment:
Based on the issue, it seems like this still means we're going to iterate over all the entries -- maybe this should be a FxHashSet or so, instead of a vec?

Reply:
So, when I try this, tests/rustdoc/intra-doc/prim-precedence.rs starts failing. I haven't quite dug in yet but I suspect it's for a good reason - I _think_ the precedence of primitive links may be relying on an ordering property of the Vec?

In any case if you want to look at the modified commit with FxHashSet it's here: https://github.com/rust-lang/rust/compare/master...jsha:rust:uniquify-intra-doc-link-2?expand=1

Reply:
Swapping FxHashSet out for FxIndexSet there seems to make tests pass

Reply:
Aha, perfect! Thanks for the tip.